src/PureSMT/SExpr.hs

```diff
@@ -186,26 +210,33 @@ ppSExpr = go 0
             . showString ")"
         List es -> showString "(" . many (map (new (n + 2)) es) . showString ")"
 
+infixr 5 :<
+
+pattern (:<) :: Char -> BS.ByteString -> BS.ByteString
+pattern c :< rest <- (BS.uncons -> Just (c, rest))
+
 -- | Parse an s-expression.
-readSExpr :: String -> Maybe (SExpr, String)
-readSExpr (c : more) | isSpace c = readSExpr more
-readSExpr (';' : more) = readSExpr $ drop 1 $ dropWhile (/= '\n') more
-readSExpr ('|' : more) = do
-  (sym, '|' : rest) <- pure (span ('|' /=) more)
-  Just (Atom ('|' : sym ++ ['|']), rest)
-readSExpr ('(' : more) = do
-  (xs, more1) <- list more
-  return (List xs, more1)
+readSExpr :: BS.ByteString -> Maybe (SExpr, BS.ByteString)
+readSExpr (c :< more) | isSpace c = readSExpr more
+readSExpr (';' :< more) = BS.dropWhile (/= '\n') more & BS.drop 1 & readSExpr
```

Comment:
Very minor thing, but I'd vote for rewriting it in the other direction with `$` — `$` is used in the other branches, and I'd say it's more common in Haskell code.

Reply:
Yeah I like `&`, `<&>` and `-.` way better but I guess that's just a me thing :'(